leader-election-impl/src/test/java/com/palantir/leader/proxy/AwaitingLeadershipProxyTest.java

```diff
@@ -67,4 +81,103 @@ public void shouldAllowObjectMethodsWhenNotLeading() throws Exception {
         assertThat(proxy.equals(null)).isFalse();
         assertThat(proxy.toString()).startsWith("com.palantir.leader.proxy.AwaitingLeadershipProxy@");
     }
+
+    @Test
+    public void shouldMapInterruptedExceptionToNCLEIfLeadingStatusChanges() throws Exception {
+        Future callToProxy = makeARequestThatThrowsAndCauseLeadershipLoss(new InterruptedException(TEST_MESSAGE));
+
+        assertThat(catchThrowable(callToProxy::get).getCause())
+                .isInstanceOf(NotCurrentLeaderException.class)
+                .hasMessage("received an interrupt due to leader election.")
+                .hasCauseExactlyInstanceOf(InterruptedException.class)
+                .hasStackTraceContaining(TEST_MESSAGE);
+    }
+
+    @Test
+    public void shouldNotMapOtherExceptionToNCLEIfLeadingStatusChanges() throws InterruptedException {
+        Future callToProxy = makeARequestThatThrowsAndCauseLeadershipLoss(new IOException(TEST_MESSAGE));
+
+        assertThat(catchThrowable(callToProxy::get).getCause())
+                .isInstanceOf(IOException.class)
+                .hasStackTraceContaining(TEST_MESSAGE);
+    }
+
+    @Test
+    public void shouldNotMapInterruptedExceptionToNCLEIfLeadingStatusDoesntChange() throws InterruptedException {
+        //Always leading
+        Callable proxy = getCallableProxy(
+                new InterruptedException(TEST_MESSAGE),
+                LeaderElectionService.StillLeadingStatus.LEADING,
+                LeaderElectionService.StillLeadingStatus.LEADING);
+
+        Future callToProxy = asyncCallToProxy(proxy);
+
+        assertThatThrownBy(proxy::call)
+                .isInstanceOf(InterruptedException.class)
+                .hasMessage(TEST_MESSAGE);
+
+        assertThat(catchThrowable(callToProxy::get).getCause())
+                .isInstanceOf(InterruptedException.class)
+                .hasMessage(TEST_MESSAGE);
+    }
+
+    private Future makeARequestThatThrowsAndCauseLeadershipLoss(Exception ex) throws InterruptedException {
+        //leadership lost while request in flight
+        Callable proxy = getCallableProxy(
+                ex,
+                LeaderElectionService.StillLeadingStatus.LEADING,
+                LeaderElectionService.StillLeadingStatus.NOT_LEADING);
+
+        Future callToProxy = asyncCallToProxy(proxy);
+
+        assertThatThrownBy(proxy::call).isInstanceOf(NotCurrentLeaderException.class)
+                .hasMessage("method invoked on a non-leader (leadership lost)");
+        return callToProxy;
+    }
+
+    private Callable getCallableProxy(
+            Exception ex,
+            LeaderElectionService.StillLeadingStatus status1,
+            LeaderElectionService.StillLeadingStatus status2) throws InterruptedException {
+
+        Callable runnableWithInterruptedException = () -> {
+            Uninterruptibles.sleepUninterruptibly(10, TimeUnit.SECONDS);
```

Comment:
One strategy that might be useful here is using a `CountdownLatch`. You can do:
```
CountdownLatch latch = new CountdownLatch(1);
Callable callable = () -> {
    latch.await();
    throw ...;
}
```

this could maybe be used as:
```
Callable proxy = proxyFor(() -> {
    latch.await();
    throw ...;
}
Future blockingCall = executor.submit(() -> proxy.call());

loseLeadership();
latch.countdown();

assertThatThrownBy(blockingCall.get()).isInstanceOf(...);
```